### PR TITLE
DRAFT: Post Middleware

### LINKED ISSUE #337: Swiping to upvote should have the upvote apply instantly, not after network roundtrip
Just like what happens when we press the upvote button, where it displays and feeds back instantly, we should have something similar with swiping and letting go.

### PR DESCRIPTION
**Pull request by EricBAndrews**
# Checklist
- [x] I have read [CONTRIBUTING.md](./CONTRIBUTING.md)
- [x] I have described what this PR contains
- [x] This PR addresses one or more open issues that were assigned to me:
      - Progress on #73 
      - Closes #337 
- [x] If this PR alters the UI, I have attached pictures/videos

# Pull Request Information

This PR introduces middleware between the raw API and the internal models for posts:
- Replaces all uses of `APIPostView` downstream of the `PostRepository` with `PostModel`. `PostRepository` calls `APIClient`, reformats the response into a `PostModel`, and passes that downstream
- Updates `PostTracker` to sit on top of `PostRepository` instead of the raw API, and deal entirely in `PostModel`
- Moves all post-related calls into `PostRepository`

The new `PostTracker` additionally takes on middleware responsibilities:
- Offers methods to perform interactions on tracked posts
- Performs state faking at the model level. Interaction methods listed above immediately update the tracker with a state-faked model, then override that with the API response whenever that comes back--if the prediction was correct, this shouldn't cause a view redraw
- Calls haptics

# Regression Testing

I have performed the following regression tests:

- For all posts: test in large, headline, and compact mode
- [x] Upvoting a post
  - [x] From feed with button
  - [x] From feed with swipe
  - [x] From expanded post
- [x] Downvoting a post
  - [x] From feed with button
  - [x] From feed with swipe
  - [x] From expanded post
- [x] Saving a post
  - [x] From feed with button
  - [x] From feed with swipe
  - [x] From expanded post
- [x] Replying to a post
  - [x] From feed with button
  - [x] From feed with swipe
  - [x] From expanded post
- [x] Reporting a post
  - [x] From feed
  - [x] From expanded post
- [x] Replying to a post
  - [x] From feed with button
  - [x] From feed with swipe
  - [x] From expanded post
- [x] Infinite feed loading
- [x] Graceful end-of-feed loading
- [x] Removing posts when blocking user
- [x] Removing posts when blocking community
- [x] Navigating to a post
  - [x] From feed
  - [x] From inbox
  - [x] From profile
- [x] Composing a new post
- [x] Editing a post
  - [x] From feed
  - [x] From expanded post

### CODE REVIEW

**Comment by Sjmarf** [comment]
Nice!

For me, swipe-to-upvote is slower in the inbox than the main feed.

**Comment by EricBAndrews** [comment]
Profile tab is going to be broken until it gets a tracker update of its own--I've opened a follow-up issue for that, but the bad behavior is a knock-on effect of moving state faking out of the post view.

I've fixed the corner radius thing--it was causing some titles to get clipped (e.g., titles starting with j). Now the corner radius is conditional on expand status, so it should work properly in both modes.

File: Mlem/API/APIClient/APIClient+Post.swift [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+            sort: sort,
+            type: type,
+            limit: limit,
+            savedOnly: savedOnly,
+            communityName: communityName
+        )
+        
+        return try await perform(request: request)
+            .posts
+            .map { PostModel(from: $0) }
```
[the remark this answers is not in the document]
Comment (EricBAndrews) [reply Mlem/API/APIClient/APIClient+Post.swift:33]: Yep, not sure how that one got through 😶

File: Mlem/Views/Shared/Posts/Lazy Load Expanded Post.swift [hunk 2]
```diff
@@ -9,24 +9,23 @@ import Dependencies
 import SwiftUI
 
 /*
- A simple wrapper around ExpandedPost which loads the
- APIPostView on demand for when we don't already have one
+ A simple wrapper around ExpandedPost which loads the PostModel on demand for when we don't already have one
  */
 struct LazyLoadExpandedPost: View {
-    @Dependency(\.apiClient) var apiClient
+    @Dependency(\.postRepository) var postRepository
```
[the remark this answers is not in the document]
Comment (EricBAndrews) [reply Mlem/Views/Shared/Posts/Lazy Load Expanded Post.swift:15]: Yep, nice catch

File: Mlem/Models/Trackers/Post Tracker New.swift [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+    }
+    
+    /**
+     Loads a single post and adds it to the tracker
+     
+     - Parameters:
+        - postId: id of the post to load
+     - Returns:
+        - PostModel of the newly loaded post
+     */
```
[the remark this answers is not in the document]
Comment (EricBAndrews) [reply Mlem/Models/Trackers/Post Tracker New.swift]: OK, I've removed every single instance of `/**` from the codebase 🙃

**Pull request merged**